File: docs/scripting/samples/csharp-scripts/examples/Event_Add_Plate_Step.cs
/**
 * Event Add Plate Step
 * 
 * Example script showing basic Cellario protocol automation and customization features
 */

#region Header
// ** Copyright HighRes Biosolutions, Inc. 2020 **
// 
//    File:      EventScript_AddPlateStep.cs 
//    Project:   ScriptLibrary
//    Solution:  CellarioWPF
#endregion

using System;
using System.Linq;
using HRB.Cellario.Scripting.API;

// TODO - customize namespace and class name
namespace HRB.Cellario.Scripting.ScriptLibrary
{
    public class EventScript_AddPlateStep : AbstractScript
    {
        /// <summary>
        /// Method called at start of Cellario run.
        /// </summary>
        /// <remarks>Expects a simple protocol with one thread with a Dispense Operation already existing</remarks>
        /// <param name="api">Access to properties and methods for interacting with the Cellario run-time scheduler, 
        /// samples, resources and devices operations.</param>
        public override void Execute(IScriptingApi api)
        {
            //about this script 
            var runorderId = api.CurrentRun.RunNumber;
            var runOrderDesc = api.CurrentRun.Description;
            var protocolName = api.CurrentRun.Protocol.ProtocolName;

            var message = "Before Order State Change for run = " + runorderId.ToString() + " \n";
            message += "Order Description '" + runOrderDesc + "' \n";
            message += "Order State - " + api.CurrentRun.RunState + " \n";
            message += "Protocol Name - " + protocolName + " \n";
            api.Messaging.Notify("Before Order State Change", message);

            TimeSpan addSimTime = TimeSpan.FromSeconds(180);

            var protocol = api.CurrentRun.Protocol;
            var threads = protocol.Threads;
            var steps = threads.First().Steps;
            steps.First().SimulationTime = addSimTime;

            var plate = api.GetPlates().FirstOrDefault();
            var stepCount = plate.RemainingSteps.Count();
            var cloneStep = plate.RemainingSteps.First(a => a.StepName == "Dispense").CloneStep();
            var cloneMoveStep = plate.RemainingSteps.First(a => a.StepName == "Move").CloneStep();
            plate.RemainingSteps.InsertStep(cloneMoveStep);
            plate.RemainingSteps.InsertStep(cloneStep);

            api.Messaging.Notify("Add Step", "added step to first plate");
        }
    }
}

File: docs/scripting/samples/python-scripts/python-examples/Event_Add_Plate_Step.py
"""
Event Add Plate Step

Python scripting example demonstrating Cellario API integration and automation capabilities
"""

#Cellario modules
from HRB.Cellario.Scripting import *
from HRB.Cellario.Scripting.API import *
import clr
clr.AddReference("System")
from System import *


# Method called during Cellario protocol execution when a sample arrives at the scripting step.
# <remarks>Expects a simple protocol with one thread with a Dispense Operation already existing</remarks>
# <Object name="api">Access to properties and methods for interacting with the Cellario run-time scheduler, 
# samples, resources and devices operations.</param>
def Execute(api : PythonScriptingApi):
    run_order_id = api.CurrentPlate.CurrentRun.RunNumber
    run_order_desc = api.CurrentPlate.CurrentRun.Description
    protocol_name = api.CurrentPlate.CurrentProtocol.ProtocolName
    
    message = ("Before Order State Change for run = " + str(run_order_id) + " \n"
              + "Order Description '" + str(run_order_desc) + "' \n"
              + "Protocol Name - " + protocol_name + " \n")
    api.Messaging.Notify("Before Order State Change", message)
    
    new_sim_time = TimeSpan(0, 0, 45)
    
    protocol = api.CurrentPlate.CurrentProtocol
    threads = protocol.Threads
    steps = threads[0].Steps
    #api.Messaging.Notify("Steps ", str(steps))
    if steps[0] == None:
        api.Messaging.Notify("Step[0] ", "doesn't exist")
    else:
        steps[0].SimulationTime = new_sim_time
        plates_list = api.GetPlates()
        plate = next(e for e in plates_list)
        #step_count = plate.RemainingSteps.len()
        clone_step = next(e for e in plate.RemainingSteps if e.StepName == 'Dispense').CloneStep()
        api.Messaging.Notify("Steps ", "Dispense step is cloned")
        clone_move_step = next(e for e in plate.RemainingSteps if e.StepName == 'Move').CloneStep()
        api.Messaging.Notify("Steps ", "Move step is cloned")
        plate.RemainingSteps.InsertStep(clone_move_step)
        plate.RemainingSteps.InsertStep(clone_step)
        
        api.Messaging.Notify("Add Step", "added step to first plate")